File: framework/languages/dotnet/src/Zlink.HttpClient/Runtime/HttpRequestSpec.cs
/* SPDX-License-Identifier: Apache-2.0 */

namespace Zlink.HttpClient.Runtime;

/// <summary>
///     Internal description of a single request as assembled by <see cref="ZLinkHttpRequestBuilder" />.
///     Mirrors the C++ <c>http_request_t</c>. <see cref="Target" /> is the path (starting with
///     <c>/</c>) with percent-encoded query already appended.
/// </summary>
internal sealed class HttpRequestSpec
{
    public required ZLinkHttpMethod Method { get; init; }

    public required string Target { get; init; }

    public byte[]? Body { get; init; }

    public Func<byte[]?>? BodyProvider { get; init; }

    public required IReadOnlyDictionary<string, string> Headers { get; init; }

    public TimeSpan? Timeout { get; init; }

    public Action<ReadOnlyMemory<byte>>? Sink { get; init; }

    public bool IsStreaming => Sink is not null || BodyProvider is not null;
}

File: framework/languages/java/zlink-http-client/src/main/java/systems/zlink/httpclient/internal/HttpRequestSpec.java
/* SPDX-License-Identifier: Apache-2.0 */
package systems.zlink.httpclient.internal;

import java.time.Duration;
import java.util.Map;
import java.util.function.Consumer;
import java.util.function.Supplier;
import systems.zlink.httpclient.ZLinkHttpMethod;

/**
 * Internal description of a single request as assembled by the request builder. Mirrors the C++
 * {@code http_request_t}. {@code target} is the path (starting with {@code /}) with percent-encoded
 * query already appended.
 */
public record HttpRequestSpec(
    ZLinkHttpMethod method,
    String target,
    String body,
    Supplier<byte[]> bodyProvider,
    Map<String, String> headers,
    Duration timeout,
    Consumer<byte[]> sink) {

    public boolean isStreaming() {
        return sink != null || bodyProvider != null;
    }
}
